### PR TITLE
Get shard from PK field for postgres generated IDs

### PR DESCRIPTION
**Pull request by 4word**
So a lot of Django magic (and by association, many modules for Django like rest framework) is possible because of automatic primary key lookups, for example, in your URLs being able to do things like `("url/<pk>/" blah)` and Django / rest framework knowing how to handle it.

Unfortunately, with sharding, that is no longer possible. Its such a big pain that I can physically not bear the thought of overwriting so much magic to make things like primary key based detail lookups possible.

So, this PR introduces a function in the router for extracting the shard from a PostgresShardGeneratedIDField. Since the postgres shards are generated by a stored procedure that inserts the shard ID into specific bits, those bits can be extracted and converted back into an id, which can then be compared to the database settings and matched up.

This will reintroduce functionality lost by going to a sharded architecture for things like detail views in DRF and generic class based views in django.

I am not sure if unit tests will pass the first time, this was a stab based on how we are handling shard extraction at work (we extract the shard and then just use the using() statement, which I want to move away from if it can be automatic)

### CODE REVIEW

**Comment by 4word** [comment]
Don't merge this in, I think I just realized that the unit tests are not valid at all because the test I am running is using a model that isn't actually sharded. I might have to add a a new model for testing this, which I am not sure I can get to this week.

**Comment by JBKahn** [comment]
k I'll take a look on saturday since tomorrow night I'm busy. In which case I'll pull this and the other PR (after I fix it) into this release.

File: tests/test_router.py [hunk 1]
```diff
@@ -441,3 +445,31 @@ def test_lookup_fallback_if_migration_directory_not_the_same_as_the_model(self):
             can_migrate_default=True,
             can_migrate_shard=False,
         )
+
+
+class RouterForPostgresIDFieldTest(TransactionTestCase):
+
+    def setUp(self):
+        from django.contrib.auth import get_user_model
```
Comment (JBKahn) [inline tests/test_router.py]: why the import?

Comment (4word) [inline tests/test_router.py]: Forgot to remove it lol. I was originally using the main user model but it was setup to use the default shard group, so I created a new one and left that there on accident. Will remove it tomorrow.

Comment (4word) [inline tests/test_router.py]: Meant to fix Friday but got busy. Fixing now.

**Comment by JBKahn** [comment]
I'll take a look asap.

File: django_sharding_library/router.py [hunk 2]
```diff
@@ -22,10 +22,16 @@ def get_specific_database_or_none(self, model):
...
+        try:
+            return instance._state.db or instance.get_shard()
+        except:
+            return None
 
     def get_shard_for_id_field(self, model, sharded_by_field_id):
-        return model.get_shard_from_id(sharded_by_field_id)
+        try:
+            return model.get_shard_from_id(sharded_by_field_id)
+        except:
```
Comment (JBKahn) [inline django_sharding_library/router.py]: Did you find cases that required this?

File: django_sharding_library/router.py [hunk 3]
```diff
@@ -22,10 +22,16 @@ def get_specific_database_or_none(self, model):
         return getattr(model, 'django_sharding__database', None)
 
     def get_shard_for_instance(self, instance):
-        return instance._state.db or instance.get_shard()
+        try:
+            return instance._state.db or instance.get_shard()
+        except:
```
Comment (JBKahn) [inline django_sharding_library/router.py]: ditto. I don't love the idea of a generic try/except unless there's a good reason for it. Did you hit anything specific?

**Comment by JBKahn** [comment]
Can you add a test case or two for the last two changes?

**Comment by JBKahn** [comment]
tests and pull master :)

File: django_sharding_library/router.py (hunk 2 above)
Comment (4word) [inline django_sharding_library/router.py]: Actually the old code just excepted and then tried the same thing again (not sure why) so I just removed the badly handled exception.

File: django_sharding_library/router.py (hunk 3 above)
Comment (4word) [inline django_sharding_library/router.py]: No, just removed it. Not sure why it was there to begin with

**Comment by 4word** [comment]
I will write up a unit test for the positional arguments. All the other changes have tests that hit them already.

**Comment by JBKahn** [comment]
Thanks, then we can get this in and do one more round of tests and then I'm good to release.

**Comment by 4word** [comment]
Merging in master was a pain with the router. I added in some more verbose comment to explain the logic for the _get_shard function(s). Will get a unit test next.

File: django_sharding_library/router.py [hunk 4]
```diff
@@ -11,27 +16,62 @@ class ShardedRouter(object):
     the wrong database as well as infer, when possible, which database to
     read or write from.
     """
-    def get_shard_for_instance(self, instance):
-        return instance._state.db or instance.get_shard()
+
+    def get_shard_for_id_field(self, model, sharded_by_field_id):
```
Comment (JBKahn) [inline django_sharding_library/router.py]: is this used anywhere?

File: django_sharding_library/router.py [hunk 5]
```diff
@@ -11,27 +16,62 @@ class ShardedRouter(object):
...
                 getattr(model, 'django_sharding__sharded_by_field'), None
             )
             if sharded_by_field_id:
-                return model.get_shard_from_id(sharded_by_field_id)
-
-        if hints.get("instance", None):
-            return get_database_for_model_instance(instance=hints["instance"])
-
-        return None
+                shard = model.get_shard_from_id(sharded_by_field_id)
```
Comment (JBKahn) [inline django_sharding_library/router.py]: this should be `model_has_sharded_id_field`

Comment (4word) [inline django_sharding_library/router.py]: Yep I am fixing it now

File: django_sharding_library/router.py [hunk 6]
```diff
@@ -11,27 +16,62 @@ class ShardedRouter(object):
...
+        if shard is None and isinstance(getattr(model._meta, 'pk'), PostgresShardGeneratedIDField) and \
```
Comment (JBKahn) [inline django_sharding_library/router.py]:
```pyton
is_pk_postgres_generated_id_field = isinstance(getattr(model._meta, 'pk'), PostgresShardGeneratedIDField)
lookup_pk = hints.get('exact_lookups', {}).get('pk') or hints.get('exact_lookups', {}).get('id')

if shard is None and is_pk_postgres_generated_id_field and lookup_pk != None:
    return self.get_shard_for_postgres_pk_field(model, lookup_pk)
```

File: tests/test_fields.py [hunk 7]
```diff
@@ -179,6 +180,10 @@ def test_pre_save_calls_save_shard(self):
 
 class PostgresShardIdFieldTestCase(TestCase):
 
+    def setUp(self):
+        from django.contrib.auth import get_user_model
```
Comment (JBKahn) [inline tests/test_fields.py]: not used

File: django_sharding_library/router.py [hunk 8]
```diff
@@ -11,27 +16,62 @@ class ShardedRouter(object):
...
+            # example is a .get(), but someone could do anything) and if it excepts, we want to fall back to the next
+            # attempt at grabbing a shard based on other potential filters
+            return None
+
+    def get_shard_for_postgres_pk_field(self, model, pk_value):
+        group = getattr(model, 'django_sharding__shard_group', None)
+        shard_id_to_find = int(bin(pk_value)[-23:-10], 2)  # We know where the shard id is stored in the PK's bits.
+
+        # We can check the shard id from the PK against the shard ID in the databases config
+        for alias in settings.DATABASES.keys():
```
Comment (JBKahn) [inline django_sharding_library/router.py]:
```python
for alias, db_settings in settings.DATABASES.items():
    if db_settings["SHARD_GROUP"] == group and db_settings["SHARD_ID"] == shard_id_to_find:
        return alias
```

File: runtests.py [hunk 9]
```diff
@@ -45,6 +45,18 @@
             'environment_variable': 'SHARD_002_DATABASE_URL',
             'default_database_url': 'sqlite://testing127'
         },
+        {
+            'name': 'app_shard_003',
+            'shard_group': 'postgres',
+            'environment_variable': 'SHARD_003_DATABASE_URL',
+            'default_database_url': 'postgres://postgres:@localhost/shard_003' if TRAVISCI else 'sqlite://testing128'
```
Comment (JBKahn) [inline runtests.py]: remove these ifs.

**Comment by JBKahn** [comment]
Update: https://github.com/JBKahn/django-sharding/blob/f0c0eb9de3a99b2198b00553166c00ec6ac55af1/tests/test_travis.py

and while you're there add an assertion about the number of DBs.

**Comment by JBKahn** [comment]
If you want I can pick up these fixes?

**Comment by JBKahn** [comment]
I'm going to merge and fix this up tonight, after that I'll test it on our systems and if all is good I'll probably do a release and we can address anything you find with another release.

**Pull request merged**